### PR TITLE
feat: optionally install Scenario with `ops[testing]` and expose the names in ops.testing

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
Add a new optional install `testing`, e.g. `pip install ops[testing]`. This pulls in a compatible version of `ops-scenario`, and exposes the Scenario names in the `ops.testing` namespace, alongside Harness.

`pip install ops[harness]` is also supported to ease the (far off, presumably 3.0) transition to Harness being moved out of the base install. It currently installs no extra dependencies, so is the same as `pip install ops` but a forward-looking charm would use `pip install ops[harness]` (or `pip install ops[harness, testing]`) if using Harness.

Requires ops-scenario 7.0.5, which has the required adjustments to support insertion into `ops.testing`.

The `ActionFailed` name exists in both `ops._private.harness` and `scenario.context`. This is handled by adjusting the Harness class to support the functionality of both and monkeypatching that into Scenario until Scenario starts using it. It's compatible with both Harness and Scenario, but will have empty data in an attribute (which attribute depends on which framework is used).

The `Container` name in `ops.testing`, which is only present for backwards compatibility, is also overwritten if ops-scenario is installed. If anyone is actually using `ops.testing.Container` instead of `ops.Container` then they'll need to fix their code before using `ops[testing]` (or ops-scenario combined with the release of ops with this change).

A very basic unit test is added to make sure that Scenario tests work (all the actual Scenario tests are in the ops-scenario repo) if ops-scenario/ops[testing] is installed (this is the case for `tox -e unit`). A test is also added to ensure that all of the Scenario names are documented, since `automodule` isn't used any more.

Also adjusts the documentation to include the new framework in ops.testing.

### CODE REVIEW

**Comment by tonyandrewmeyer** [comment]
@PietroPasotti @benhoyt @dimaqq I imagine this PR will have a bit of back-and-forth before it's completely done. Would you please give it an initial look-over? Probably focusing on the main approaches rather than specifics for now, and then once everyone is happy with that there can be a second review for approval that goes deeper. Thanks!

**Comment by tonyandrewmeyer** [comment]
(I'll look into the failing test with Python 3.10 - I hadn't checked different Pythons locally, so will do that).

**Comment by dimaqq** [comment]
CI fail: `TypeError: EventBase.__init__() got an unexpected keyword argument 'app'`

3.12 was not run / did not get to complete because 3.10 failed.

**Comment by tonyandrewmeyer** [comment]
> CI fail: `TypeError: EventBase.__init__() got an unexpected keyword argument 'app'`
> 
> 3.12 was not run / did not get to complete because 3.10 failed.

Yeah, the tests all run locally for me with 3.8, 3.9, 3.10, 3.11, 3.12. Still looking into this, but it seems unlikely that it needs to block a high-level review.

File: docs/requirements.txt
Comment (dimaqq) [inline docs/requirements.txt:76]: 🤔 does ops belong here? I'd imagine that docs are generated from the current repo.

Does it hurt? e.g. if someone is working on a branch, could this pull in different version of code?

File: ops/testing.py
Comment (dimaqq) [inline ops/testing.py:52]: 🤔 don't we want to hide the name "scenario" from like everywhere?

**Comment by dimaqq** [comment]
A point for discussion about the overall approach:
- do we want to keep separate pypi:ops-scenario going forward?
- or do shall we vendor scenario code and rename things?

File: docs/requirements.txt
Comment (tonyandrewmeyer) [reply docs/requirements.txt:76]: > 🤔 does ops belong here? I'd imagine that docs are generated from the current repo.

This file is auto-generated. ops ends up here because there's a circle: `ops-scenario` depends on `ops` (that could possibly be removed in time, but it would be a bunch of refactoring work) and `ops[testing]` depends on `ops-scenario`. The docs need to have `ops-scenario` installed to be able to pull in all the docstrings, and installing `ops-scenario` installs `ops`.

> Does it hurt? e.g. if someone is working on a branch, could this pull in different version of code?

Good question! It seems to work ok - I edited one of the ops modules locally and did `tox -e docs` and it picked up my local changes. I'm not totally sure about when readthedocs does the build - the config has:

```yaml
python:
  install:
    - method: pip
      path: .
      extra_requirements:
        - docs
```

So it's ignoring docs/requirements.txt I believe. I'm not sure what would happen if a branch of `ops` had a version number incompatible with the requirements in `ops-scenario` - probably an error? But that does seem like to be something we can manage ourselves, especially once we've merged the repos.

File: ops/testing.py
Comment (tonyandrewmeyer) [reply ops/testing.py:52]: > 🤔 don't we want to hide the name "scenario" from like everywhere?

I think we want (or are expected to) hide it so that it's not used in charms, in our docs, etc. Essentially: you can build charms and write tests for them without ever knowing the name "Scenario".

It's not 100% hidden, e.g. you can do this:

```python
>>> repr(ops.testing.Context)
"<class 'scenario.context.Context'>"
```

Although arguably `Context.__repr__` could be set to something better. But there's still things like:

```python
>> ops.testing.State.__module__
'scenario.state'
```

I don't feel that the "I want to reserve the name Scenario" comment means that we need to entirely wipe all traces of it.

**Comment by tonyandrewmeyer** [comment]
> * do we want to keep separate pypi:ops-scenario going forward?
> * or do shall we vendor scenario code and rename things?

We have to have a separate package in order to (cleanly) have one set of code that gets used when building charms and an augmented set of code that gets used when testing charms. I very strongly feel that the test framework should not be in the bundled charm. I think the correct way to do that is to have a package for testing (eventually two: one for Harness and one for Scenario) rather than things like charmcraft ripping out the folder when packing.

We are intending to add the contents of the ops-scenario repository to the operator repository, but my preference is to do that as a separate PR after this one. We would then publish two packages from the operator repository. I still feel that's going to be a better maintenance burden than having two repos. I don't really see a benefit in renaming things as part of this, but that can be discussed when we get to it.

We could publish Scenario under a new name, like `ops-testing`. I don't see much value in that either, though, and I like having a small nod to the history of how this framework came to be.

**Comment by tonyandrewmeyer** [comment]
I can reproduce the failing test locally if I turn pytest-xdist off (probably other small numbers work too). I assume some other test is leaking into the one that's failing, so it only fails when both of those tests are executed in the same runner.

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:55]: This reads a bit odd, like a subject is missing. I think we should say "State-transition tests allow you to ..." or something.

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:58]: There's a lot of text after this point before the developer can see anything in action. I think it'd be great to show a tiny example right here:

```python
def test_start():
    ctx = Context(MyCharm)
    state = State(...something small here...)
    out = ctx.run(ctx.on.start(), state)
    assert isinstance(out.unit_status, ActiveStatus)
```

Maybe we want *something* really simple in the `State` so readers can see that.

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:67]:
```suggestion
These 'state-transition' tests give charm authors a way to test
how the state changes in reaction to events. They are not
necessarily tests of individual methods or functions;
they are testing the 'contract' of the charm: given
a certain state, when a certain event happens, the charm should transition to
another state. Unlike integration tests, they do not test using a real Juju
controller and model, and focus on a single Juju unit.
For simplicity, we refer to them as 'unit' tests.
```

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:75]:
```suggestion
Writing these tests should nudge you into thinking of a charm as a black-box
'input to output' function. The inputs are:

- Event: why am I, the charm, being executed
- State: am I the leader? what is my integration data? what is my config?
- Context: what integrations can I have? what containers can I have?

The output is another `State`: the state after
the charm has interacted with the mocked Juju model.
The output state is the same type of data structure as the input state.
```

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:83]: I would reverse those two bullet points as the "output state" one is by far the most important.

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:92]: "run the context" sounds odd. Maybe this could just be one bullet point with "run a specific Juju event"?

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:94]:
```suggestion
    - verify that the output state is what you expect it to be
```

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:96]: Maybe drop this bullet point altogether? We don't need to mention this more advanced feature right here.

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:99]: The wording here is a bad odd/casual. But maybe if we start with a tiny example above, we don't need this paragraph or example at all?

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:110]: It's a pity we have to list these out manually? Is there any way to avoid that, so that we don't forget to add new classes here when we add them to the code?

**Review by benhoyt — COMMENTED** [review]
Adding some further comments on the docs.

File: ops/charm.py
Comment (tonyandrewmeyer) [reply ops/charm.py:244]: Scenario copies various docstrings from ops objects, particularly the statuses and the events. When there's something like `:class:ConfigChangedEvent` in the Scenario objects and Sphinx does `autoclass`, it can't find the `ConfigChangedEvent` object, so fails. With the leading `ops.` Sphinx knows where to find them even when the docstring is being used outside of ops. The `~` tells Sphinx to just display the name and not the `ops.`.

The docstring copying code could try to be smarter and auto-fix these (it's probably just replacing ":class:" with "class:ops." (and the backticks), with maybe some false positives to work around), or Scenario could have its own docstrings (maybe just saying "go read the ops class for more details").

Discussed this in a call and Ben's ok with leaving them as-is for the moment, now that it's clearer what they are there for.

File: ops/testing.py
Comment (tonyandrewmeyer) [reply ops/testing.py:19]: It was because I couldn't get Sphinx to properly work with `__all__` when using `automodule`. However, I had to give up using `automodule` anyway, because I couldn't find a clean way to get it to use `ops.testing` rather than `ops._private.harness` and `scenario` for references.

So that means that `__all__` is actually viable - I'll go back to that. I'll also rely on `__all__` defining the API and not bother importing `typing as _typing` and so on.

File: ops/testing.py
Comment (tonyandrewmeyer) [reply ops/testing.py:93]: I think it was accidentally not included in the `scenario` top level: https://github.com/canonical/ops-scenario/pull/200

File: docs/index.rst
Comment (tonyandrewmeyer) [reply docs/index.rst:110]: I do agree. I couldn't get `automodule` to work and also use the `ops.testing` name rather than leaking `ops._private.harness` or `scenario`. It's somewhat working against the `automodule` intention since the usual desired behaviour is to document anything defined in the module but not anything that the module pulls in to use - but here we have basically nothing defined in the module, and want to document (almost) everything that's pulled in.

There could be some combination of Sphinx configuration options that make it work, but I wasn't able to find them with the limited amount of time I allowed for it. It's also possible that we could create our own `automodule` directive (subclassing or patching the original one, probably) that would solve our issues - but this seemed like it would be a lot more work.

For now, I've added a test_infra test to make it less likely that we forget to maintain this, and a comment about why we do this, and hopefully someone in the future will make this nicer.

**Comment by tonyandrewmeyer** [comment]
@tmihoc the code changes won't be interesting to you, but you might like to review the doc changes? It's probably easiest to look over the [generated docs](https://ops--1381.org.readthedocs.build/en/1381/) since it pulls in the definitions from ops-scenario as well (which you already reviewed, of course).

File: docs/index.rst
Comment (tonyandrewmeyer) [reply docs/index.rst]: I've split it out - you might want to take another look now?

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:83]: Do we need `_version = None` here, so the `if _version` below doesn't raise a `NameError`?

File: docs/harness.rst
Comment (benhoyt) [inline docs/harness.rst:10]:
```suggestion
    install in an ops release in the future. Charm authors that don't want to upgrade
    will still be able to use it with ``pip install ops[harness]``.
```

File: docs/state-transition-testing.rst
Comment (benhoyt) [reply docs/state-transition-testing.rst:5]: Tony and I roughly agree on changing the menu to this:

* API reference
* Unit testing (was "Scenario")
* Harness (legacy unit testing)

Perhaps also say `State-transition tests (previously known as "Scenario") ...` in the page content.

**Review by benhoyt — COMMENTED** [review]
Overall looking great, just comments on doc that we've discussed.

File: ops/_private/harness.py
Comment (dimaqq) [inline ops/_private/harness.py:202]: This would be the same, wouldn't it?

```suggestion
        self.output = output or ActionOutput([], {})
```

File: ops/__init__.py
Comment (dimaqq) [inline ops/__init__.py:17]: I'm pretty sure "core features" was intended...

**Review by dimaqq — APPROVED** [review]

File: ops/testing.py
Comment (tonyandrewmeyer) [reply ops/testing.py:83]: No, because the `if _version` is in the try's `else` block, so if the exception is raised it won't get to there.

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:16]:
```suggestion
    recommended approach (it was previously known as 'Scenario').
```

File: docs/index.rst
Comment (benhoyt) [inline docs/index.rst:19]:
```suggestion
    moved out of the base ``ops`` install in an ops release in the future; at that point, charm
```

Comment (benhoyt) [reply docs/index.rst:19]: It also seems like too much information for this point in the docs. Can we move everything from "It will be moved out..." to the Harness page?

**Review by benhoyt — APPROVED** [review]
Approving! Left a couple of minor take-or-leave comments.

File: docs/index.rst
Comment (tonyandrewmeyer) [reply docs/index.rst:19]: Happy to move it - in terms of "at that point": I updated the description to include this but didn't mention it in a comment: I added the ability to do `pip install ops[harness]` now (this was in the spec, but I had forgotten). This eases transition in the future, because charms can use `pip install ops[harness]` now and it should seamlessly work when it becomes separated.

**Pull request merged**
